sql/src/main/java/com/vmware/ddlog/DDlogJooqProvider.java

```diff
@@ -24,78 +42,226 @@
+            final String tableName = ((com.facebook.presto.sql.tree.Table) node.getFrom().get()).getName().toString();
+            final List<Field<?>> fields = tables.get(tableName.toUpperCase());
+            if (fields == null) {
+                throw new RuntimeException(String.format("Unknown table %s queried in statement: %s", tableName, sql));
+            }
+            final Result<Record> result = dslContext.newResult(fields);
+            try {
+                dDlogAPI.dumpTable(ddlogRelationName(tableName), (record, l) -> {
+                    final Record jooqRecord = dslContext.newRecord(fields);
+                    final Object[] returnValue = new Object[fields.size()];
+                    for (int i = 0; i < fields.size(); i++) {
+                        returnValue[i] = structToValue(fields.get(i), record.getStructField(i));
+                    }
+                    jooqRecord.fromArray(returnValue);
+                    result.add(jooqRecord);
+                });
+            } catch (final DDlogException e) {
+                throw new RuntimeException(e);
+            }
+            return new MockResult(1, result);
         }
-        final Result<Record> result = dslContext.newResult(fields);
-        try {
-            dDlogAPI.dumpTable("R" + tableName.toLowerCase(), (record, l) -> {
-                final Record jooqRecord = dslContext.newRecord(fields);
-                final Object[] returnValue = new Object[fields.size()];
-                for (int i = 0; i < fields.size(); i++) {
-                    returnValue[i] = structToValue(fields.get(i), record.getStructField(i));
+
+        @Override
+        protected MockResult visitQuery(final Query node, final String context) {
+            final QuerySpecification specification = (QuerySpecification) node.getQueryBody();
+            return visitQuerySpecification(specification, context);
+        }
+
+        @Override
+        protected MockResult visitInsert(final Insert node, final String sql) {
+            try {
+                // The assertions below encode assumption A2 (see javadoc for the DDlogJooqProvider class)
+                assert node.getQuery().getQueryBody() instanceof Values;
+                final Values values = (Values) node.getQuery().getQueryBody();
+                final String tableName = node.getTarget().toString();
+                final List<Field<?>> fields = tables.get(tableName.toUpperCase());
+                final int tableId = dDlogAPI.getTableId(ddlogRelationName(tableName));
+                for (final Expression row: values.getRows()) {
+                    assert row instanceof Row;
+                    final List<Expression> items = ((Row) row).getItems();
+                    assert items.size() == fields.size();
+                    final DDlogRecord[] recordsArray = new DDlogRecord[items.size()];
+                    for (int i = 0; i < items.size(); i++) {
```

Comment:
I have some helper function Linq.map/zip/any that you may want to occasionally use.

Reply:
I could use Java streams too but this looks simple enough.